doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,810 @@
+`DTLSv1_get_timeout()`. Calling `SSL_tick()` on an `SSL` object will enable it
+to process any outstanding time based events. For a DTLS application this is
+exactly equivalent to calling `DTLSv1_handle_timeout()`. In a QUIC application
+only the `SSL` object for the connection would ever have time based events. For
+a TLS application there would never be time based events (so calling
+`SSL_tick()` will have no effect).
+
+## Changes to the BIO API
+
+In order to support QUIC the current `BIO` API will need to be extended. The
+existing "dgram" `BIO` can be used for UDP. However this only has very limited
+support for sockets operating in an "unconnected" state. Currently, sending a
+UDP packet from a dgram `BIO` using an unconnected socket requires the caller to
+first use `BIO_ctrl()` to set the destination address and then call `BIO_write()`
+to actually send the packet. A similar process is required for obtaining the peer
+address from a received packet. This multi-call approach is unlikely to scale
+well in a multi threaded application.
+
+A further problem is the existing API provides no mechanism for sending or
+receiving multiple messages at the same time. Operating systems provide various
+mechanisms for optimising sending/receiving data if this can be achieved. Note
+that OpenSSL does not need to implement support for this in its own `BIO`s in its
+initial releases - but adding support for it in the API enables others to do so,
+and enables it to be added easily to OpenSSL's own `BIO`s at any point. Initially
+OpenSSL's own dgram BIO can simply convert requests to send multiple messages to
+sequentially sending/receiving single messages.
+
+The existing API also provides no support for setting/obtaining the local
+address for a message to be sent/received. This may be useful in the case where
+a socket is bound to multiple address using `INADDR_ANY` or `IN6ADDR_ANY`.
+
+To solve all of the above issues two new functions are proposed: `BIO_sendmmsg()`
+and `BIO_recvmmsg()`. There will also be associated `BIO_METHOD` functions to set
+implementations for these. Support can be added to fall back to the existing
+sending/receiving approach in the event that a `BIO` does not support the new
+`BIO_METHOD` functions.
+
+A demonstration client using the new SSL API functions has been provided in
+[Appendix B](#appendix-b).
+
+# <a id='appendix-a'></a>Appendix A: SSL API ssl.h and bio.h additions
+
+Proposed additions to the ssl.h header file are as follows:
+
+
+````
+/* Methods representing the QUIC protocol */
+__owur const SSL_METHOD *QUIC_method(void);
+__owur const SSL_METHOD *QUIC_server_method(void);
+__owur const SSL_METHOD *QUIC_client_method(void);
+
+/*
+ * An SSL object represents a single stream. The "first" SSL object for a
+ * connection additionally represents that connection. In QUIC multiple SSL
+ * objects (streams) can be associated with the same connection id. In (D)TLS
```

Comment:
Can the stream associated with the connection object be destroyed before the connection is (other live streams are still active)?

Reply:
Yeah, why not? I suppose SSL_read/write would then just fail on the object.